torchgen/dest/register_dispatch_key.py

```diff
@@ -409,6 +416,77 @@ def generate_defn(cpp_sig: CppSignature) -> str:
                 metadata = self.backend_index.get_kernel(f)
                 if metadata is None:
                     return None
+
+                # define the struct for dispatch-less composite kernels.
+                impl_struct = ""
+                if self.backend_index.should_gen_dispatchless_composite(f):
+                    # define the native function as a wrapper static method, redirecting the call to
+                    # the appropriate function (e.g. on CPU, on CUDA, or as a CompositeExplicitAutograd).
+                    def dispatchless_function_defn(
+                        f: NativeFunction, g: Optional[NativeFunctionsGroup]
+                    ) -> str:
+                        dispatch_key = self.backend_index.dispatch_key
+
+                        # the declaration signature is a static method (not a Tensor method).
+                        decl_sig = CppSignatureGroup.from_native_function(
+                            f, method=False, fallback_binding=False
+                        ).signature
+
+                        # decide 2 things:
+                        #     1. where is the redirect function? (i.e. its namespace)
+                        #     2. is it a Tensor method? (some operations are only available as methods)
+                        if self.backend_index.has_registered_kernel(f, g):
+                            prefix = f"at::{dispatch_key.lower()}::"
+                            is_method = False
+                        else:
+                            # give preference to function variants instead of method.
+                            is_method = {Variant.method} == f.variants
+                            if Variant.function in f.variants:
+                                if (
+                                    dispatch_key
+                                    == DispatchKey.CompositeExplicitAutograd
+                                ):
+                                    prefix = "at::"
+                                else:
+                                    # dependent dispatch-less composite kernel headers are not included
+                                    # for CPU or CUDA. However, we shouldn't hit this point, since they
+                                    # should have a dispatch-less kernel generated at this dispatch key.
+                                    prefix = "at::native::"
```

Comment:
Shouldn't we raise an error then?

From my understanding, I think that it should only be valid to generate one of these composite kernels if all of the listed OperatorNames have at least one dispatch key entry from one of the keys in `COMPOSITE_DISPATCH_KEYS`.

If an op happens to have both a CPU and a CompositeExplicitAutograd entry then we should prefer the direct CPU one.

That means that the call that you generate will always be one of:
```
at::cpu::foo
at::cuda::foo
at::meta::foo
at::compositeexplicitautograd::foo
```

Will that help to simplify some of this logic? (no need to worry about the tensor method case, or the at::native case)

Reply:
Right. I guess that does make sense. It's also clearer on how these calls are generated.

Reply:
I guess we can also include `at::compositeimplicitautograd:foo`, too, right!?

Reply:
Although we can ignore the `at::native` case, I don't think we can do the same with the tensor method case. 

For example, say a kernel makes use of `resize_`. While CPU gets generated without problem (it has a CPU dispatch), CompositeExplicitAutograd can't help but rely on the tensor method call. 

It's either that, or not generating a CompositeExplicitAutograd at all. I particularly think it makes sense to generate the slow-path calls (i.e. supporting tensor methods and `at::`). What do you think?

Reply:
yep, totally agreed with you now that I'm looking at the updated PR